File: 0910. Smallest Range II/Solution.cpp
#include <algorithm>
#include <ranges>
#include <vector>
using namespace std;

class Solution {
public:
    int smallestRangeII(vector<int>& nums, int k)
    {
        ranges::sort(nums);
        int ans = nums.back() - nums[0];
        for (size_t i = 1; i < nums.size(); ++i) {
            int mx = max(nums[i - 1] + k, nums.back() - k);
            int mn = min(nums[0] + k, nums[i] - k);
            ans = min(ans, mx - mn);
        }
        return ans;
    }
};

File: 0910. Smallest Range II/Solution.py
from typing import List


class Solution:
    def smallestRangeII(self, nums: List[int], k: int) -> int:
        nums.sort()
        ans = nums[-1] - nums[0]
        for i in range(1, len(nums)):
            mx = max(nums[i - 1] + k, nums[-1] - k)
            mn = min(nums[0] + k, nums[i] - k)
            ans = min(ans, mx - mn)
        return ans
